react/the_react_ecosystem/react_router.md

```diff
@@ -0,0 +1,589 @@
+
+~~~jsx
+import { useParams } from "react-router-dom";
+import DefaultProfile from "./DefaultProfile";
+import Spinach from "./Spinach";
+import Popeye from "./Popeye";
+
+const Profile = () => {
+  const { name } = useParams();
+
+  return (
+    <div>
+      <h1>Hello from profile page!</h1>
+      <p>So, how are you?</p>
+      <hr />
+      <h2>The profile visited is here:</h2>
+      {name === "popeye" ? (
+        <Popeye />
+      ) : name === "Spinach" ? (
+        <Spinach />
+      ) : (
+        <DefaultProfile />
+      )}
+    </div>
+  );
+};
+
+export default Profile;
+~~~
+
+### Handling Bad Urls
+
+But alas, the index path doesn't work with this anymore, as in the `/profile` path, no params are actually passed. Actually, the `/profile` path doesn't make much sense without an actual name, else whose profile is it supposed to show, right? So, the application shows an error! This can't be good, so how do you show a default page in case the user visits a wrong or unused path? You can pass in an `errorElement` argument here! Create a basic "Not Found" page:
+
+~~~jsx
+import { Link } from "react-router-dom";
+
+const ErrorPage = () => {
+  return (
+    <div>
+      <h1>Oh no, this route doesn't exist!</h1>
+      <Link to="/">
+        You can go back to the home page by clicking here, though!
+      </Link>
+    </div>
+  );
+};
+
+export default ErrorPage;
+~~~
+
+Add in the `errorElement` to the configuration, and check it out! The `/profile`, or any unmentioned paths render an error page!
+
+~~~jsx
+import React from "react";
+import ReactDOM from "react-dom/client";
+import { createBrowserRouter, RouterProvider } from "react-router-dom";
+import App from "./App";
+import Profile from "./Profile";
+import ErrorPage from "./ErrorPage";
+
+const router = createBrowserRouter([
+  {
+    path: "/",
+    element: <App />,
+    errorElement: <ErrorPage />,
+  },
+  {
+    path: "profile/:name",
+    element: <Profile />,
+  },
+]);
+
+ReactDOM.createRoot(document.getElementById("root")).render(
+  <React.StrictMode>
+    <RouterProvider router={router} />
+  </React.StrictMode>
+);
+~~~
+
+### Protected Routes
+
+Often, you will face a need to decide when a certain route is rendered or not. One use case is authentication, where you may want to render certain routes based on if the user is logged in or not. While there are many ways to do so, one of the easiest ways can be to conditionally creating a config for the router.
+
+But before we do that, let us refactor our routes to a component of their own, so that we can add whatever conditional logic we want, if it exists as a hook (remember, we can't use hooks outside of a React component!). Even if you don't have any need for a conditional rendering of routes, it is much neater nonetheless, to have them seperate. Create a new `Router.js` component and move your routes to it.
+
```

Comment:
Really liking the section on protected routes, but I feel like we're giving out too much code. I don't find a need to hand out implementation for protected routes for the learners here, rather letting them using external resources and figuring it out on their own. We can expand on the text explanation and add an example of a website that implements a protected route. What do you think?

Reply:
I don't exactly know of a website that implements this kind of protected routes, but I can look for it. Mostly, I came up with it for this new docs.

Reply:
From some searching, these came up as a decent couple of examples:
- https://stackoverflow.com/questions/62384395/protected-route-with-react-router-v6
  - This is mostly a way to do the protected routes and probably one of the better answers I could find. 
  - The first answer is similar to the example here, the second one and onward though, are older answers.
- https://github.com/iammanishshrma/react-protected-routes/tree/master/src/routes
  - Yet another example, this one implemented with a much granular `ProtectedRoute` component, but yes, this works.
  - This uses Redux though, which needs to be a disclaimer. The reader may see some custom hooks here.

On top of these, maybe we could refactor this to be a __Refactoring the router__ section, and a different __Protected Routes__ section?

Reply:
We could add another example in the form of a stackblitz thing.

Reply:
I'm saying we should discuss protected routes as a concept. An example of a protected route would be: https://www.theodinproject.com/dashboard as it can only be accessed when a user is signed in. We don't need to give out code or any specific code implementation here, learners can figure it out themselves by researching and playing around. I hope that makes sense.

What do you have in mind for Refactoring the router section?

Reply:
Removed the code provided in the Protected Routes section, and added it a simple explanation, while adding a couple of implementations as an example in the additional resources section.